File: public/javascripts/custom_info.js
/**
 Copyright 2013 Red Hat, Inc.

 This software is licensed to you under the GNU General Public
 License as published by the Free Software Foundation; either version
 2 of the License (GPLv2) or (at your option) any later version.
 There is NO WARRANTY for this software, express or implied,
 including the implied warranties of MERCHANTABILITY,
 NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
 have received a copy of GPLv2 along with this software; if not, see
 http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
*/

var KT = (KT === undefined) ? {} : KT;

KT.custom_info = (function() {

    $("#new_custom_info_keyname").live("keyup", function() {
        check_for_empty($(this));
    });

    $(".remove_custom_info_button").live('click', function(e) {
        e.preventDefault();
        remove_custom_info($(this));
    });

    $("#create_custom_info_button").live('click', function(e) {
        e.preventDefault();
        create_custom_info($(this));
    });

    function check_for_empty($textfield) {
        $button = $("#create_custom_info_button");
        if ($textfield.val().length > 0 ) {
            $button.removeAttr("disabled");
        } else {
            $button.attr("disabled", "true");
        }
    }

    function remove_custom_info($button) {
        $.ajax({
            url    : $button.data("url"),
            type   : $button.data("method"),
            success: function() {
                remove_custom_info_row($button.data("id"));
            },
            error  : function(data) {
                notices.displayNotice("error", window.JSON.stringify({ "notices": [$.parseJSON(data.responseText)["displayMessage"]] }));
            }
        });
    }

    function create_custom_info($button) {
        var keyname = $("#new_custom_info_keyname").val();
        var value = $("#new_custom_info_value").val();

        $.ajax({
            url     : $button.data("url"),
            type    : $button.data("method"),
            dataType: 'json',
            data    : { "keyname": keyname, "value": value },
            success : function(data) {
                add_custom_info_row(data);
            },
            error   : function(data) {
                notices.displayNotice("error", window.JSON.stringify({ "notices": [$.parseJSON(data.responseText)["displayMessage"]] }));
            }
        });
    }

    function remove_custom_info_row(data_id) {
        $("tr[data-id='" + data_id + "']").remove();
    }

    function add_custom_info_row(data) {
        var esc_keyname = escape(data["keyname"]);
        var _keyname = data["keyname"].replace(" ", "_");
        var value = data["value"];
        var informable_type = data["informable_type"];
        var informable_id = data["informable_id"];
        var update_path = KT.routes.api_update_custom_info_path(informable_type, informable_id, esc_keyname);
        var destroy_path = KT.routes.api_destroy_custom_info_path(informable_type, informable_id, esc_keyname);

        var new_row = "<tr class=\"primary_color\" data-id=\"custom_info_" + _keyname + "\">"
        + "<td class=\"ra\">"
        + "<label for=\"custom_info_" + _keyname + "\">" + data["keyname"] + "</label>"
        + "</td>"
        + "<td>"
        + "<div class=\"editable edit_textfield\" data-method=\"put\" data-url=\"" + update_path + "\" name=\"value\" style title=\"Click to edit\">" + value + "</div>"
        + "</td>"
        + "<td>"
        + "<input class=\"btn warning remove_custom_info_button\" data-id=\"custom_info_" + _keyname + "\" data-method=\"delete\" data-url=\"" + destroy_path + "\" type=\"submit\" value=\"remove\">"
        + "</td>"
        + "</tr>";

        $("#new_custom_info_row").after(new_row);
        setTimeout(function() { $("tr[data-id='custom_info_" + _keyname + "']").addClass("row_fade_in"); }, 1);
        $("#new_custom_info_keyname").val("");
        $("#new_custom_info_value").val("");
        check_for_empty($("#new_custom_info_keyname"));

        var $new_editable = $("tr[data-id='custom_info_" + _keyname + "']").find(".editable");
        var common_settings = {
            method     :  'PUT',
            cancel     :  i18n.cancel,
            submit     :  i18n.save,
            indicator  :  i18n.saving,
            tooltip    :  i18n.clickToEdit,
            placeholder:  i18n.clickToEdit,
            submitdata :  $.extend({ authenticity_token: AUTH_TOKEN }, KT.common.getSearchParams()),
            onerror    :  function(settings, original, xhr) {
                original.reset();
                $("#notification").replaceWith(xhr.responseText);
                notices.checkNotices();
            }
        };
        var settings = {
            type :  'text',
            width:  270,
            name :  $new_editable.attr('name')
        };
        $new_editable.editable($new_editable.attr('data-url'), $.extend(common_settings, settings));
    }

})();

File: app/models/katello/custom_info.rb
#
# Copyright 2014 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

module Katello
class CustomInfo < Katello::Model
  self.include_root_in_json = false

  attr_accessible :keyname, :value, :org_default

  belongs_to :informable, :polymorphic => true

  validates :keyname, :presence => true, :length => { :maximum => 255 },
                      :uniqueness => {:scope => [:informable_type, :informable_id],
                                      :message => "already exists for this object"}
  validates :value, :length => { :maximum => 255 }

  validates :informable_id, :presence => true
  validates :informable_type, :presence => true

  before_validation :strip_attributes
  after_save :reindex_informable

  def self.find_by_informable_keyname(informable, keyname)
    return informable.custom_info.find_by_keyname(keyname)
  end

  # find the Katello object by type and ID (i.e. "system", 32)
  def self.find_informable(informable_type, informable_id)
    class_name = informable_type.classify
    class_name = "Katello::#{class_name}" unless class_name.start_with?("Katello::")
    informable = class_name.constantize.find(informable_id)
    fail _("Resource %s does not support custom information") % class_name unless informable.respond_to? :custom_info
    return informable
  end

  def reindex_informable
    self.informable.class.index.import([self.informable])
  end

  def to_s
    "#{self.keyname}: #{self.value.nil? ? _("NOT-SPECIFIED") : self.value}"
  end

  def <=>(other)
    return self.keyname <=> other.keyname
  end

  private

  def strip_attributes
    self.keyname.try(:strip!)
    self.value.try(:strip!)
  end

end
end
